ppci/lang/c/semantics.py

```diff
@@ -174,19 +174,30 @@ def on_variable_initialization(self, variable, expression):
         if self.scope.is_definition(variable.name):
             self.error("Invalid redefinition.", variable.location)
 
-        self.patch_size_from_initializer(variable.typ, expression)
+        variable.typ = self.patch_size_from_initializer(
+            variable.typ,
+            expression
+        )
         variable.initial_value = expression
 
     def patch_size_from_initializer(self, typ, initializer):
-        """ Fill array size from elements! """
-
+        """ Fill array type size from elements if the type size is unknown.
+            Return a new type if modified 
+        """
         if typ.is_array and typ.size is None:
             if isinstance(initializer, expressions.ArrayInitializer):
-                typ.size = len(initializer.values)
+                new_type = types.ArrayType(
+                    typ.element_type,
+                    len(initializer.values)
+                ) 
+                initializer.typ = new_type
```

Comment:
I think we should not modify the type of the initializer here, right? It should already be good.